### PR TITLE
Convert to Typescript

### PR DESCRIPTION
**Pull request by bwrrp**
Also move to using mocha for tests, as nothing in this library is browser-specific.

Build configurations are included for AMD, commonjs and ES modules, as well as single-file UMD and ESM bundles.

### CODE REVIEW

File: src/Node.ts [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+	 * Node.importNode, as with Node.cloneNode() and Node.renameNode() operations (though Node.adoptNode does
+	 * preserve the information), and equality tests in Node.isEqualNode() do not consider user data in making the
+	 * assessment.
+	 *
+	 * This method offers the convenience of associating data with specific nodes without needing to alter the
+	 * structure of a document and in a standard fashion, but it also means that extra steps may need to be taken
+	 * if one wishes to serialize the information or include the information upon clone, import, or rename
+	 * operations.
+	 *
+	 * @param key Key under which the value is stored
```
Comment (DrRataplan) [inline src/Node.ts]: missing `@param` for data

File: src/Document.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	/**
+	 * (internal) The ranges that are active on the current document.
+	 */
+	public _ranges: Range[] = [];
+
+	constructor () {
+		super(Node.DOCUMENT_NODE);
+
+		// Non-standard: should be null for Document nodes.
+		this.ownerDocument = this;
```
Comment (bwrrp) [inline src/Document.ts]: Will fix this in 2.0

**Comment by DrRataplan** [comment]
Don't forget to update the readme!

File: src/DocumentType.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	 */
+	constructor (name: string, publicId: string, systemId: string) {
+		super(Node.DOCUMENT_TYPE_NODE);
+
+		this.name = name;
+		this.publicId = publicId;
+		this.systemId = systemId;
+	}
+
+	public cloneNode (deep: boolean = true, _copy?: DocumentType): DocumentType {
```
Comment (bwrrp) [inline src/DocumentType.ts]: The underscore was intended to indicate private. I agree that it clashes in a confusing way with the convention for unused arguments. I'll remove the underscores for function arguments and just document them as non-standard.

File: src/Node.ts (hunk 1 above)
Comment (bwrrp) [inline src/Node.ts]: I think I got all of them now

**Comment by DrRataplan** [comment]
Looks great! :shipit:

**Comment by timschlechter** [comment]
👍 😁

**Pull request merged**